### PR TITLE
Refactor Mesh Quality Kernels into .cxx

### PR DESCRIPTION
**Pull request by nimberledge**
Refactored all .cxx kernels into .cxx files, ensured all relevant tests passed (even those for decomposition in test_decomposition.py).

Use ufl.sqrt and ufl.as_vector in metric.py to ensure the decomposition tests all passed.

### CODE REVIEW

**Review by jwallwork23 — CHANGES_REQUESTED** [review]
Thanks a lot for this! I just have a couple of requests:
1. Could you add brief comments to serve as docstrings for the C++ kernels? (See [here](https://github.com/pyroteus/pyroteus/blob/c3f9ea7886f59cf8ded90771e3edf0e2a69f02bf/pyroteus/cxx/eigen_kernels2d.cxx#L34-L37), for example.)
2. `pyroteus/kernels.py` is now kind of redundant. Would it make sense to move `EigenKernelHandler` to `pyroteus/metric.py` and `QualityKernelHandler` to `pyroteus/quality`?

**Comment by jwallwork23** [comment]
I ran `test_decomposition.py` and `test_quality.py` locally and they passed.

**Comment by jwallwork23** [comment]
Also, I guess `eigen_kernel` would need to be merged into the KernelHandler classes somehow.

**Comment by acse-akm121** [comment]
Changes are now made, tests have been re-run.

**Comment by nimberledge** [comment]
Changes are now made, tests have been re-run

File: pyroteus/metric.py [hunk 1]
```diff
@@ -18,6 +17,49 @@
...
+    def __call__(self, d):
+        """
+        :arg d: the spatial dimension
+        """
+        assert d in (2, 3), f"Spatial dimension {d} not supported."
+        eigen_kernels = os.path.join(os.path.dirname(__file__), "cxx/metric{:d}d.cxx")
+        return open(eigen_kernels.format(d)).read()
+
+    @staticmethod
+    def get_pyop_kernel(kernel, *args, **kwargs):
```
Comment (jwallwork23) [inline pyroteus/metric.py:41]: Why `get_pyop_kernel` and not `get_pyop2_kernel`?

File: pyroteus/quality.py [hunk 2]
```diff
@@ -1,10 +1,68 @@
...
+
+    def __call__(self, d):
+        """
+        :arg d: the spatial dimension
+        """
+        assert d in (2, 3), f"Spatial dimension {d} not supported."
+        if self.d_restrict is not None:
+            assert d in self.d_restrict, f"Spatial dimension {d} not supported for {self.__name__}"
+
+        qual_kernels = os.path.join(os.path.dirname(__file__), "cxx/quality{}d.cxx".format(d))
```
Comment (jwallwork23) [inline pyroteus/quality.py:40]: Please could we use an f-string?

File: pyroteus/metric.py [hunk 3]
```diff
@@ -18,6 +17,49 @@
            "determine_metric_complexity", "density_and_quotients", "check_spd"]
 
 
+class EigenKernelHandler():
```
Comment (jwallwork23) [inline pyroteus/metric.py:20]: Just for consistency, could we rename this to `MetricKernelHandler`?

**Review by jwallwork23 — CHANGES_REQUESTED** [review]
Looks great. Just three minor comments.

**Review by jwallwork23 — APPROVED** [review]

**Comment by jwallwork23** [comment]
Thanks, this is neat :)

**Pull request merged**